profiling/src/lib.rs

```diff
@@ -0,0 +1,719 @@
+    let version = getenv(CStr::from_bytes_with_nul_unchecked(b"DD_VERSION\0"));
+    let agent_host = getenv(CStr::from_bytes_with_nul_unchecked(b"DD_AGENT_HOST\0"));
+    let trace_agent_port = getenv(CStr::from_bytes_with_nul_unchecked(
+        b"DD_TRACE_AGENT_PORT\0",
+    ));
+    let trace_agent_url = getenv(CStr::from_bytes_with_nul_unchecked(b"DD_TRACE_AGENT_URL\0"));
+
+    globals.profiling_enabled = profiling_enabled.and_then(parse_boolean).unwrap_or(false);
+
+    globals.profiling_experimental_cpu_time_enabled = profiling_experimental_cpu_time_enabled
+        .and_then(parse_boolean)
+        .unwrap_or(false);
+
+    let profiling_log_level = profiling_log_level
+        .and_then(|string| {
+            let string = string.to_ascii_uppercase();
+            LevelFilter::from_str(string.as_str()).ok()
+        })
+        .unwrap_or(LevelFilter::Off);
+
+    globals.profiling_log_level = std::mem::transmute(profiling_log_level);
+    globals.env = intern(&env);
+    globals.service = intern(&service);
+    globals.version = intern(&version);
+    globals.agent_host = intern(&agent_host);
+    globals.trace_agent_port = intern(&trace_agent_port);
+    globals.trace_agent_url = intern(&trace_agent_url);
+}
+
+pub struct RequestLocals {
+    pub last_wall_time: Instant,
+    pub last_cpu_time: Option<cpu_time::ThreadTime>,
+    pub tags: Vec<Tag>,
+    pub uri: String,
+}
+
+fn static_tags() -> Vec<Tag> {
+    vec![
+        Tag::from_value("language:php").expect("static tags to be valid"),
+        // Safety: calling getpid() is safe.
+        Tag::new("process_id", unsafe { libc::getpid() }.to_string())
+            .expect("static tags to be valid"),
+    ]
+}
+
+thread_local! {
+    static REQUEST_LOCALS: RefCell<RequestLocals> = RefCell::new(RequestLocals {
+        last_wall_time: Instant::now(),
+        last_cpu_time: None,
+        tags: static_tags(),
+        uri: String::from("http://localhost:8126"),
+    });
+}
+
+/* If Failure is returned the VM will do a C exit; try hard to avoid that,
+ * using it for catastrophic errors only.
+ */
+extern "C" fn rinit(r#type: c_int, module_number: c_int) -> ZendResult {
+    #[cfg(debug_assertions)]
+    trace!("RINIT({}, {})", r#type, module_number);
+
+    unsafe { zend::datadog_php_profiling_rinit() };
+
+    // SAFETY: globals pointer is valid during rinit.
+    let globals = unsafe { zend::datadog_php_profiling_globals_get() };
+    unsafe { read_env(globals) };
+
+    // At the moment, logging is truly global, so init it exactly once whether
+    // profiling is enabled or not.
+    static ONCE: Once = Once::new();
+    ONCE.call_once(|| {
+        #[cfg(not(debug_assertions))]
+        logging::log_init(unsafe { std::mem::transmute(globals.profiling_log_level) });
+        #[cfg(debug_assertions)]
+        log::set_max_level(globals.profiling_log_level);
+
+        unsafe {
+            // reminder: this cannot be done in minit because of Apache forking model
```

Comment:
I'm not sure what this is meant to be informing me of? It can't be done but is done?

Reply:
In the PHP extension and module ecosystem, the "MINIT" phase is when a `zend_module_entry` is initialized. You can read more about this here if you care to: https://www.phpinternalsbook.com/php7/extensions_design/php_lifecycle.html.

The comment is there to remind me that because of the way Apache forks processes, I cannot spawn a thread (which the profiler does) during MINIT. There are more comments about this on the docs for the `minit` function.